File: agentmail-negotiation-agent/typescript/src/deal.ts
/**
 * JSON-backed deal + counterparty state.
 */

import * as fs from "node:fs";

const DEAL_FILE = "deal.json";

export interface CounterpartyOffer {
  price: number;
  currency: string;
  terms_summary: string;
  meets_must_haves: boolean;
  notes: string;
  received_at?: string;
}

export interface Counterparty {
  email: string;
  name: string;
  status?: "queued" | "contacted" | "offered" | "declined" | "countered" | "target_hit" | "walked";
  thread_id?: string;
  contacted_at?: string;
  last_counter_at?: string;
  last_anchor?: number;
  decline_reason?: string;
  current_offer?: CounterpartyOffer | null;
}

export interface Deal {
  what: string;
  must_haves: string[];
  ideal_price: number;
  max_price: number;
  currency: string;
  deal_context?: string;
  counterparties: Counterparty[];
  buyer_accept_intent?: any[];
}

export function load(): Deal | null {
  if (!fs.existsSync(DEAL_FILE)) return null;
  try {
    return JSON.parse(fs.readFileSync(DEAL_FILE, "utf8")) as Deal;
  } catch {
    return null;
  }
}

export function save(state: Deal): void {
  fs.writeFileSync(DEAL_FILE, JSON.stringify(state, null, 2));
}

export function getCounterpartyByEmail(email: string): Counterparty | undefined {
  const deal = load();
  if (!deal) return undefined;
  return deal.counterparties.find((cp) => cp.email.toLowerCase() === email.toLowerCase());
}

export function getCounterpartyByThread(threadId: string): Counterparty | undefined {
  const deal = load();
  if (!deal) return undefined;
  return deal.counterparties.find((cp) => cp.thread_id === threadId);
}

export function updateCounterparty(
  email: string,
  fields: Partial<Counterparty>,
): Counterparty | undefined {
  const deal = load();
  if (!deal) return undefined;
  let target: Counterparty | undefined;
  for (const cp of deal.counterparties) {
    if (cp.email.toLowerCase() === email.toLowerCase()) {
      Object.assign(cp, fields);
      target = cp;
      break;
    }
  }
  if (target) save(deal);
  return target;
}

export function queuedCounterparties(): Counterparty[] {
  const deal = load();
  if (!deal) return [];
  return deal.counterparties.filter(
    (cp) => !cp.status || cp.status === "queued",
  );
}

export function allReplied(d?: Deal | null): boolean {
  const deal = d ?? load();
  if (!deal || !deal.counterparties.length) return false;
  const terminal = new Set(["offered", "declined", "walked", "target_hit"]);
  return deal.counterparties.every((cp) => terminal.has(cp.status || ""));
}

File: agentmail-negotiation-agent/python/deal.py
"""
JSON-backed deal state.

`deal.json` schema:
  {
    "what": "...",                      # what's being negotiated
    "must_haves": ["...", "..."],
    "ideal_price": 36000,
    "max_price": 39500,
    "currency": "USD",
    "deal_context": "...",              # free-text extra info (trade-in, financing, deadlines)
    "counterparties": [
      {
        "email": "...",
        "name": "...",
        "status": "queued|contacted|offered|declined|countered|target_hit|walked",
        "thread_id": "...",
        "current_offer": {              # populated by record_offer
          "price": 35000,
          "currency": "USD",
          "terms_summary": "...",
          "meets_must_haves": true,
          "notes": "..."
        }
      }
    ]
  }

`buyer_email` and `buyer_thread_id` get added at runtime when the agent first
contacts the buyer (so we can route the round summary back to the same thread).
"""

import json
from pathlib import Path

DEAL_FILE = Path("deal.json")


def load() -> dict:
    if not DEAL_FILE.exists():
        return {}
    try:
        return json.loads(DEAL_FILE.read_text())
    except Exception:
        return {}


def save(state: dict) -> None:
    DEAL_FILE.write_text(json.dumps(state, indent=2))


def get_counterparty_by_email(email: str) -> dict | None:
    deal = load()
    for cp in deal.get("counterparties", []):
        if cp.get("email", "").lower() == email.lower():
            return cp
    return None


def get_counterparty_by_thread(thread_id: str) -> dict | None:
    deal = load()
    for cp in deal.get("counterparties", []):
        if cp.get("thread_id") == thread_id:
            return cp
    return None


def update_counterparty(email: str, **fields) -> dict | None:
    deal = load()
    target = None
    for cp in deal.get("counterparties", []):
        if cp.get("email", "").lower() == email.lower():
            cp.update(fields)
            target = cp
            break
    if target:
        save(deal)
    return target


def queued_counterparties() -> list[dict]:
    deal = load()
    return [
        cp for cp in deal.get("counterparties", [])
        if cp.get("status", "queued") in ("", "queued")
    ]


def all_replied(deal_state: dict | None = None) -> bool:
    """True if every counterparty has either offered, declined, or walked."""
    deal = deal_state or load()
    cps = deal.get("counterparties", [])
    if not cps:
        return False
    terminal = {"offered", "declined", "walked", "target_hit"}
    return all(cp.get("status") in terminal for cp in cps)
